consensus/src/miner.rs

```diff
@@ -15,62 +15,57 @@
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
 use crate::{error::ConsensusError, Consensus};
+use snarkos_storage::{DynStorage, SerialBlock, SerialBlockHeader, SerialRecord, SerialTransaction};
 use snarkvm_algorithms::CRH;
-use snarkvm_dpc::{
-    testnet1::{instantiated::*, Record as DPCRecord},
-    AccountAddress,
-    Block,
-    BlockHeader,
-    DPCScheme,
-    RecordScheme,
-    Storage,
-    TransactionScheme,
-    Transactions as DPCTransactions,
-};
+use snarkvm_dpc::{testnet1::instantiated::*, AccountAddress, BlockHeader, BlockHeaderHash, DPCComponents};
 use snarkvm_posw::{txids_to_roots, PoswMarlin};
 use snarkvm_utilities::{bytes::ToBytes, to_bytes};
 
 use chrono::Utc;
-use rand::{thread_rng, Rng};
+use rand::thread_rng;
 use std::sync::Arc;
 
+lazy_static::lazy_static! {
+    /// The mining instance that is initialized with a proving key.
+    static ref POSW: PoswMarlin = {
+        PoswMarlin::load().expect("could not instantiate the miner")
+    };
+}
+
 /// Compiles transactions into blocks to be submitted to the network.
 /// Uses a proof of work based algorithm to find valid blocks.
-pub struct Miner<S: Storage> {
+pub struct MineContext {
     /// The coinbase address that mining rewards are assigned to.
     address: AccountAddress<Components>,
     /// The sync parameters for the network of this miner.
-    pub consensus: Arc<Consensus<S>>,
-    /// The mining instance that is initialized with a proving key.
-    miner: PoswMarlin,
+    pub consensus: Arc<Consensus>,
+    block_height: usize,
+    canon_header: SerialBlockHeader,
 }
 
-impl<S: Storage> Miner<S> {
-    /// Creates a new instance of `Miner`.
-    pub fn new(address: AccountAddress<Components>, consensus: Arc<Consensus<S>>) -> Self {
-        Self {
-            address,
-            consensus,
-            // Load the miner with the proving key, this should never fail
-            miner: PoswMarlin::load().expect("could not instantiate the miner"),
-        }
-    }
+impl MineContext {
+    pub async fn prepare(
```

Comment:
This is just `new`, no?

Reply:
`new` implies no logic, but we call into storage which isn't free.